### ISSUE TITLE
Add support for Intel XL710 (40Gbps)

### ISSUE THREAD

**Issue by atheurer** [emmericp/MoonGen#86]
I have been trying to add support for these adapters.  I use a lot of these (12 x 40Gb) in order to test for NFV workloads.  We currently use pktgen-dpdk for this testing, but I'd like to move this to Moongen since it is very customizable.

I have some very basic functionality working.  For example, I can use ./examples/benchmark/packet-sizes-rx.lua, but the Tx stats are wrong (Rx seems to be OK).  I have not looked in to time-stamping yet.

I am wondering if anyone has code changes for XL710.  If not, I have a couple small changes I can provide.

**Comment by emmericp** [comment]
I also have two dual port XL710 NICs and I ran some basic tests for the XL710 NICs a few months ago and I think I fixed the stats (but that might have been on the dpdk2.0 branch).

I have to do some more 40 Gbit tests next week anyways for the final verison of the paper anyways, so you can expect some fixes for that soon.

**Comment by emmericp** [comment]
Update: I already fixed the link detection and I'm currently working on stats tracking (which is way more complicated on the XL710).

Timestamping shouldn't be too difficult but I don't know how to do rate control yet.

The architecture of the XL710 chip is very different from the other NICs.
The simple rate control per queue feature was removed but doing rate control should still be possible, it's just going to be way more complicated in the driver.

I need to do some refactoring as the current multi-driver support is somewhat messy.

**Comment by atheurer** [comment]
Are you referring to hardware rate control?  If so, do you believe software rate control should work without many issues?  I'd be happy to test anything you have.

**Comment by emmericp** [comment]
depends on the type of software rate control

just idling between batches will work just fine, but not a single moongen script does this because it's unreliable.

the fancy :setDelay() call needs some driver support for CRC disabling

**Comment by emmericp** [comment]
l2 filters are now working, this enables ARP


next step: timestamping which is somewhat problematic: layer 3 timestamping currently relies on flow director filters which aren't supported on the i40e in the DPDK version that we are currently using.

**Comment by emmericp** [comment]
Is there any specific feature that you need to make the switch from Pktgen-DPDK?
I think the current version at least achieves feature parity on the driver level.

I'll probably post-pone layer 3 timestamping to the DPDK 2.0 upgrade due to the FDir issue.


BTW: I've done a quick performance evaluation of the TX capabilities of the XL710 NIC.
Check out Section 5.4 of our paper: http://www.net.in.tum.de/fileadmin/bibtex/publications/papers/MoonGen_IMC2015.pdf 

TL;DR: you can expect about 21 Mpps per port. This is a hardware bottleneck, adding more queues/CPUs doesn't help.

**Comment by atheurer** [comment]
On Tue, Aug 25, 2015 at 7:05 AM, Paul Emmerich <notifications@github.com>
wrote:

> Is there any specific feature that you need to make the switch from
> Pktgen-DPDK?
> I think the current version at least achieves feature parity on the driver
> level.
>
> I'll probably post-pone layer 3 timestamping to the DPDK 2.0 upgrade due
> to the FDir issue.
>
> BTW: I've done a quick performance evaluation of the TX capabilities of
> the XL710 NIC.
> Check out Section 5.4 of our paper:
> http://www.net.in.tum.de/fileadmin/bibtex/publications/papers/MoonGen_IMC2015.pdf
>
> TL;DR: you can expect about 21 Mpps per port. This is a hardware
> bottleneck, adding more queues/CPUs doesn't help.
>
>
​All this great news.  I will test it this week.

**Comment by yjiang5** [comment]
Hi, emmericp, we are trying to use MoonGen to test l2 forward latency on XL710 NIC.  How is your progress on the time stamping? If you can share your work, that will be cool and we are happy to test it.

Thanks

**Comment by emmericp** [comment]
Please use this branch at the moment: https://github.com/PrincetonUniversity/MoonGen/tree/dpdk2.1

I'll get back to the whole DPDK 2.x and XL710 stuff once I finish up with all the papers that I need to write at the moment.

**Comment by yjiang5** [comment]
Hi, emmericp, thanks for your reply and that branch works, although I noticed some negative latency value and still investigate the reason.

--jyh

**Comment by yjiang5** [comment]
After debugging, seems one reason for the negative latency value (I'm trying the examples/l2-load-latency.lua) is because current code does not support hasTimestamp() on "lua/include/timestamping.lua" for XL710 yet. After changing that code, most of the negative latency value is gone, but I still got 4 negative value after running for about 4 hours (previously >10 for 10 seconds testing).

Thanks
--jyh

**Comment by emmericp** [comment]
merged in the dpdk2.0 branch

note that it will never be perfect with these annoying XL710 NICs because, unlike all other intel NICs, they do not save the sequence number of the timestamped packet...

**Comment by yjiang5** [comment]
Hi, emmericp, I tried the dpdk2.0 branch and failed because of followed error:

[yjiang5@otcbuild01 MoonGen20]$ sudo ./build/MoonGen examples/l2-load-latency.lua --dpdk-config=lua/dpdk-scnfv02.lua 0 1
Could not run main script: lua/include/dpdk.lua:16: bad argument #1 to 'lshift' (number expected, got cdata)
Lua error: attempt to call a nil value

And the corresponding code is below. Seems the ULL is not recognized.

mod.PKT_RX_VLAN_PKT         = bit.lshift(1ULL, 0)
mod.PKT_RX_RSS_HASH         = bit.lshift(1ULL, 1)
mod.PKT_RX_FDIR             = bit.lshift(1ULL, 2)
mod.PKT_RX_L4_CKSUM_BAD     = bit.lshift(1ULL, 3)
mod.PKT_RX_IP_CKSUM_BAD     = bit.lshift(1ULL, 4)
mod.PKT_RX_EIP_CKSUM_BAD    = bit.lshift(0ULL, 0)
mod.PKT_RX_OVERSIZE         = bit.lshift(0ULL, 0)

Thanks
--jyh

**Comment by emmericp** [comment]
you are using the wrong version of luajit.
please make sure that you cloned the submodule correctly.

**Comment by yjiang5** [comment]
Thanks for quick response. I simply do the "git submodule update --init". What's the expected version to be used?  BTW, I have another MoonGen with the master branch on the machine, not sure if that will be a problem. I hope no since the luajit is installed locally on the deps/luajit/usr/local directory.

Thanks
--jyh

**Comment by emmericp** [comment]
you can see the correct revisions here: https://github.com/emmericp/MoonGen/tree/dpdk2.0/deps

try something like submodule sync and update with --remote

**Comment by yjiang5** [comment]
Thanks for reply. If I force to use the same commitID as your link (i.e. https://github.com/LuaJIT/LuaJIT/tree/6cb38f788f58d642d1b9198aace947e00f1dc2d8 ), it works. However, seems that Commit ID is for the version 2.1 branch? Not sure I missed any steps on my previous try. I clone a new MoonGen and start everything from scratch.
